File: PointRun/util/GCHelper.m
//
//  GCHelper.m
//  CatRace
//
//  Created by Ray Wenderlich on 4/23/11.
//  Copyright 2011 __MyCompanyName__. All rights reserved.
//

#import "GCHelper.h"

@implementation GCHelper
@synthesize gameCenterAvailable;
@synthesize presentingViewController;
@synthesize match;
@synthesize delegate;
@synthesize playersDict;
@synthesize invitedPlayer;
@synthesize invite;
@synthesize mainViewController;

#pragma mark Initialization

static GCHelper *sharedHelper = nil;
+ (GCHelper *) sharedInstance {
    if (!sharedHelper) {
        sharedHelper = [[GCHelper alloc] init];
    }
    return sharedHelper;
}

- (BOOL)isGameCenterAvailable {
	Class gcClass = (NSClassFromString(@"GKLocalPlayer"));
	return gcClass != nil;
}

- (id)init {
    if ((self = [super init])) {
        gameCenterAvailable = [self isGameCenterAvailable];
        if (gameCenterAvailable) {
            NSNotificationCenter *nc = 
            [NSNotificationCenter defaultCenter];
            [nc addObserver:self selector:@selector(authenticationChanged) name:GKPlayerAuthenticationDidChangeNotificationName object:nil];
        }
    }
    return self;
}

#pragma mark Internal functions

- (void)authenticationChanged {
    if ([GKLocalPlayer localPlayer].isAuthenticated && !userAuthenticated) {
        NSLog(@"Authentication changed: player authenticated.");
        userAuthenticated = TRUE;
    } else if (![GKLocalPlayer localPlayer].isAuthenticated && userAuthenticated) {
        NSLog(@"Authentication changed: player not authenticated");
        userAuthenticated = FALSE;
    }
}

- (void)lookupPlayers {
    NSMutableArray *playerIDs = [@[] mutableCopy];
    for (GKPlayer *player in match.players) {
        [playerIDs addObject:player.playerID];
    }
    
    [GKPlayer loadPlayersForIdentifiers:playerIDs withCompletionHandler:^(NSArray *players, NSError *error) {
        if (error != nil) {
            NSLog(@"Error retrieving player info: %@", error.localizedDescription);
            matchStarted = NO;
            [delegate matchEnded:@""];
        } else {
            self.playersDict = [NSMutableDictionary dictionaryWithCapacity:players.count];
            for (GKPlayer *player in players) {
                NSLog(@"Found player: %@", player.alias);
                [playersDict setObject:player forKey:player.playerID];
            }

            matchStarted = YES;
            [[GKMatchmaker sharedMatchmaker] finishMatchmakingForMatch:self.match];
            [delegate matchStarted];
        }
    }];
}

#pragma mark User functions

- (void)authenticateLocalUser {
    if (!gameCenterAvailable) return;
    NSLog(@"Authenticating local user...");
    if ([GKLocalPlayer localPlayer].authenticated == NO) {     
        [[GKLocalPlayer localPlayer] setAuthenticateHandler:^(UIViewController *view, NSError *error) {
            if (error == nil) {
                self.authenticated = YES;
            } else {
                NSLog(@"%@", error.localizedDescription);
            }
        }];
    } else {
        NSLog(@"Already authenticated!");
    }
}

- (void)findMatchWithMinPlayers:(int)minPlayers maxPlayers:(int)maxPlayers viewController:(UIViewController *)viewController delegate:(id<GCHelperDelegate>)theDelegate {
    if (!gameCenterAvailable) return;

    matchStarted = NO;
    self.match = nil;
    self.presentingViewController = viewController;
    delegate = theDelegate;
    [presentingViewController dismissViewControllerAnimated:NO completion:nil];

    GKMatchRequest *request = [[GKMatchRequest alloc] init];
    request.minPlayers = minPlayers;     
    request.maxPlayers = maxPlayers;

    GKMatchmakerViewController *mmvc = [[GKMatchmakerViewController alloc] initWithMatchRequest:request];
    mmvc.matchmakerDelegate = self;

    [presentingViewController presentViewController:mmvc animated:YES completion:nil];
}

- (void)showAchievements:(UIViewController *)viewController {
    self.presentingViewController = viewController;
    [presentingViewController dismissViewControllerAnimated:NO completion:nil];

    GKGameCenterViewController *gcvc = [[GKGameCenterViewController alloc] init];
    gcvc.viewState = GKGameCenterViewControllerStateAchievements;
    if (gcvc != nil) {
        gcvc.gameCenterDelegate = self;
        [presentingViewController presentViewController:gcvc animated:YES completion:nil];
    }
}

- (void)showLeaderboards:(UIViewController *)viewController {
    self.presentingViewController = viewController;
    [presentingViewController dismissViewControllerAnimated:NO completion:nil];
    
    GKGameCenterViewController *gcvc = [[GKGameCenterViewController alloc] init];
    gcvc.viewState = GKGameCenterViewControllerStateLeaderboards;
    if (gcvc != nil) {
        gcvc.gameCenterDelegate = self;
        [presentingViewController presentViewController:gcvc animated:YES completion:nil];
    }
}

- (void)showGameCenter:(UIViewController *)viewController {
    self.presentingViewController = viewController;
    [presentingViewController dismissViewControllerAnimated:NO completion:nil];
    
    GKGameCenterViewController *gcvc = [[GKGameCenterViewController alloc] init];
    if (gcvc != nil) {
        gcvc.gameCenterDelegate = self;
        [presentingViewController presentViewController:gcvc animated:YES completion:nil];
    }
}

- (void)reportAchievementIdentifier:(NSString *)identifier percentComplete:(float)percent {
    GKAchievement *achievement = [[GKAchievement alloc] initWithIdentifier:identifier];
    if (achievement) {
        achievement.percentComplete = percent;
        achievement.showsCompletionBanner = YES;
        [GKAchievement reportAchievements:@[achievement] withCompletionHandler:^(NSError *error) {
            if (error != nil) {
                NSLog(@"Error in reporting achievements: %@", error);
            }
        }];
    }
}

- (void)submitTimedScore:(int)highScore {
    GKScore *score = [[GKScore alloc] initWithLeaderboardIdentifier:@"mosttimedpoints"];
    score.value = [[NSUserDefaults standardUserDefaults] integerForKey:@"mostTimedPoints"];
    [GKScore reportScores:@[score] withCompletionHandler:^(NSError *error) {
        if (error) {
            NSLog(@"%@", error.localizedDescription);
        }
    }];
}

#pragma mark GKGameCenterControllerDelegate

- (void)gameCenterViewControllerDidFinish:(GKGameCenterViewController *)gameCenterViewController {
    [presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark GKMatchmakerViewControllerDelegate

// The user has cancelled matchmaking
- (void)matchmakerViewControllerWasCancelled:(GKMatchmakerViewController *)viewController {
    [presentingViewController dismissViewControllerAnimated:YES completion:nil];
    [presentingViewController.navigationController popViewControllerAnimated:YES];
}

// Matchmaking has failed with an error
- (void)matchmakerViewController:(GKMatchmakerViewController *)viewController didFailWithError:(NSError *)error {
    [presentingViewController dismissViewControllerAnimated:YES completion:nil];
    NSLog(@"Error finding match: %@", error.localizedDescription);    
}

// A peer-to-peer match has been found, the game should start
- (void)matchmakerViewController:(GKMatchmakerViewController *)viewController didFindMatch:(GKMatch *)theMatch {
    [presentingViewController dismissViewControllerAnimated:YES completion:nil];
    self.match = theMatch;
    match.delegate = self;
    if (!matchStarted && match.expectedPlayerCount == 0) {
        NSLog(@"Ready to start match!");
        [self lookupPlayers];
    }
}

#pragma mark GKMatchDelegate

// The match received data sent from the player.
- (void)match:(GKMatch *)theMatch didReceiveData:(NSData *)data fromPlayer:(NSString *)playerID {
    if (match != theMatch) return;
    [delegate match:theMatch didReceiveData:data fromPlayer:playerID];
}

// The player state changed (eg. connected or disconnected)
- (void)match:(GKMatch *)theMatch player:(NSString *)playerID didChangeState:(GKPlayerConnectionState)state {
    if (match != theMatch) return;
    switch (state) {
        case GKPlayerStateConnected: 
            if (!matchStarted && theMatch.expectedPlayerCount == 0) [self lookupPlayers];
            break; 
        case GKPlayerStateDisconnected:
            matchStarted = NO;
            [delegate matchEnded:@""];
            self.match = nil;
            break;
        case GKPlayerStateUnknown:
            break;
    }
}

// The match was unable to connect with the player due to an error.
- (void)match:(GKMatch *)theMatch connectionWithPlayerFailed:(NSString *)playerID withError:(NSError *)error {
    if (match != theMatch) return;
    NSLog(@"Failed to connect to player with error: %@", error.localizedDescription);
    matchStarted = NO;
    [delegate matchEnded:playerID];
}

// The match was unable to be established with any players due to an error.
- (void)match:(GKMatch *)theMatch didFailWithError:(NSError *)error {
    if (match != theMatch) return;
    NSLog(@"Match failed with error: %@", error.localizedDescription);
    matchStarted = NO;
    [delegate matchEnded:@""];
}

#pragma mark GKLocalPlayerListener

- (void)player:(GKPlayer *)player didAcceptInvite:(GKInvite *)inviteToAccept {
    GKMatchmakerViewController *mmvc = [[GKMatchmakerViewController alloc] initWithInvite:inviteToAccept];
    mmvc.matchmakerDelegate = self;
    [mainViewController presentViewController:mmvc animated:YES completion:nil];
}

- (void)player:(GKPlayer *)player didRequestMatchWithOtherPlayers:(NSArray *)playersToInvite {
    
}

@end

File: PointRun/util/GCHelper.swift
//
//  GCHelper.swift
//  PointRun
//
//  Created by Jack Cook on 9/27/14.
//  Copyright (c) 2014 CosmicByte. All rights reserved.
//

import GameKit

protocol GCHelperDelegate {
    func matchStarted()
    func match(match: GKMatch, didReceiveData: NSData, fromPlayer: String)
    func matchEnded()
}

class GCHelper: NSObject, GKMatchmakerViewControllerDelegate, GKGameCenterControllerDelegate, GKMatchDelegate, GKLocalPlayerListener {
    
    var presentingViewController: UIViewController!
    var match: GKMatch!
    var delegate: GCHelperDelegate!
    var playersDict = [:]
    var invitedPlayer: GKPlayer!
    var invite: GKInvite!
    
    var matchStarted = false
    var authenticated = false
    
    class var sharedInstance: GCHelper {
        struct Static {
            static let instance = GCHelper()
        }
        return Static.instance
    }
    
    override init() {
        super.init()
        nc.addObserver(self, selector: Selector("authenticationChanged"), name: GKPlayerAuthenticationDidChangeNotificationName, object: nil)
    }
    
    // MARK: Internal functions
    
    func authenticationChanged() {
        if (GKLocalPlayer.localPlayer().authenticated && authenticated) {
            println("Authentication changed: player authenticated")
            authenticated = true
        } else {
            println("Authentication changed: player not authenticated")
            authenticated = false
        }
    }
    
    func lookupPlayers() {
        var playerIDs = [String]()
        for player in match.players {
            playerIDs.append(player.playerID)
        }
        
        GKPlayer.loadPlayersForIdentifiers(playerIDs, withCompletionHandler: { (players, error) -> Void in
            if (error != nil) {
                println("Error retrieving player info: \(error.localizedDescription)")
                self.matchStarted = false
                self.delegate.matchEnded()
            } else {
                self.playersDict = NSMutableDictionary(capacity: players.count)
                for player in players {
                    println("Found player: \(player.alias)")
                    self.playersDict.setValue(player, forKey: player.playerID)
                }
                
                self.matchStarted = true
                GKMatchmaker.sharedMatchmaker().finishMatchmakingForMatch(self.match)
                self.delegate.matchStarted()
            }
        })
    }
    
    // MARK: User functions
    
    func authenticateLocalUser() {
        println("Authenticating local user...")
        if (GKLocalPlayer.localPlayer().authenticated == false) {
            GKLocalPlayer.localPlayer().authenticateHandler = { (view, error) in
                if (error == nil) {
                    self.authenticated = true
                } else {
                    println("\(error.localizedDescription)")
                }
            }
        } else {
            println("Already authenticated")
        }
    }
    
    func findMatchWithMinPlayers(minPlayers: Int, maxPlayers: Int, viewController: UIViewController, delegate theDelegate: GCHelperDelegate) {
        matchStarted = false
        match = nil
        presentingViewController = viewController
        delegate = theDelegate
        presentingViewController.dismissViewControllerAnimated(false, completion: nil)
        
        var request = GKMatchRequest()
        request.minPlayers = minPlayers
        request.maxPlayers = maxPlayers
        
        var mmvc = GKMatchmakerViewController(matchRequest: request)
        mmvc.matchmakerDelegate = self
        
        presentingViewController.presentViewController(mmvc, animated: true, completion: nil)
    }
    
    func reportAchievementIdentifier(identifier: String, percent: Double) {
        var achievement: GKAchievement? = GKAchievement(identifier: identifier)
        if let a = achievement {
            achievement!.percentComplete = percent
            achievement!.showsCompletionBanner = true
            GKAchievement.reportAchievements([achievement!], withCompletionHandler: { (error) -> Void in
                if (error != nil) {
                    println("Error in reporting achievements: \(error)")
                }
            })
        }
    }
    
    func reportLeaderboardIdentifier(identifier: String, score: Int) {
        var scoreObject = GKScore(leaderboardIdentifier: identifier)
        scoreObject.value = Int64(score)
        println("\(scoreObject.value)")
        GKScore.reportScores([scoreObject], withCompletionHandler: { (error) -> Void in
            if (error != nil) {
                println("Error in reporting leaderboard scores: \(error)")
            }
        })
    }
    
    func showGameCenter(viewController: UIViewController, viewState: GKGameCenterViewControllerState) {
        presentingViewController = viewController
        
        var gcvc = GKGameCenterViewController()
        gcvc.viewState = viewState
        gcvc.gameCenterDelegate = self
        presentingViewController.presentViewController(gcvc, animated: true, completion: nil)
    }
    
    // MARK: GKGameCenterControllerDelegate
    
    func gameCenterViewControllerDidFinish(gameCenterViewController: GKGameCenterViewController!) {
        presentingViewController.dismissViewControllerAnimated(true, completion: nil)
    }
    
    // MARK: GKMatchmakerViewControllerDelegate
    
    func matchmakerViewControllerWasCancelled(viewController: GKMatchmakerViewController!) {
        presentingViewController.dismissViewControllerAnimated(true, completion: nil)
        presentingViewController.navigationController?.popViewControllerAnimated(true)
    }
    
    func matchmakerViewController(viewController: GKMatchmakerViewController!, didFailWithError error: NSError!) {
        presentingViewController.dismissViewControllerAnimated(true, completion: nil)
        println("Error finding match: \(error.localizedDescription)")
    }
    
    func matchmakerViewController(viewController: GKMatchmakerViewController!, didFindMatch theMatch: GKMatch!) {
        presentingViewController.dismissViewControllerAnimated(true, completion: nil)
        match = theMatch
        match.delegate = self
        if (!matchStarted && match.expectedPlayerCount == 0) {
            println("Ready to start match!")
            self.lookupPlayers()
        }
    }
    
    // MARK: GKMatchDelegate
    
    func match(theMatch: GKMatch!, didReceiveData data: NSData!, fromPlayer playerID: String!) {
        if (match != theMatch) {
            return
        }
        
        delegate.match(theMatch, didReceiveData: data, fromPlayer: playerID)
    }
    
    func match(theMatch: GKMatch!, player playerID: String!, didChangeState state: GKPlayerConnectionState) {
        if (match != theMatch) {
            return
        }
        
        switch (state) {
        case GKPlayerConnectionState.StateConnected:
            if (!matchStarted && theMatch.expectedPlayerCount == 0) {
                lookupPlayers()
            }
        case GKPlayerConnectionState.StateDisconnected:
            matchStarted = false
            delegate.matchEnded()
            match = nil
        default:
            break
        }
    }
    
    func match(theMatch: GKMatch!, didFailWithError error: NSError!) {
        if (match != theMatch) {
            return
        }
        
        println("Match failed with error: \(error.localizedDescription)")
        matchStarted = false
        delegate.matchEnded()
    }
    
    // MARK: GKLocalPlayerListener
    
    func player(player: GKPlayer!, didAcceptInvite inviteToAccept: GKInvite!) {
        var mmvc = GKMatchmakerViewController(invite: inviteToAccept)
        mmvc.matchmakerDelegate = self
        presentingViewController.presentViewController(mmvc, animated: true, completion: nil)
    }
    
    func player(player: GKPlayer!, didRequestMatchWithOtherPlayers playersToInvite: [AnyObject]!) {
        
    }
}